### PR TITLE
Added reporting slowest tests

### PR DESCRIPTION
**Pull request by Radinyn**

### CODE REVIEW

File: protostar/commands/test/slowest_test_formatting.py
Comment (kasperski95) [inline protostar/commands/test/slowest_test_formatting.py:35]: This could be a part of `TestingSummary` class.

File: protostar/commands/test/test_command.py
Comment (kasperski95) [inline protostar/commands/test/test_command.py:211]: We need to keep `TestCommand` from growing. If you move the logic to the `TestingSummary` you can pass slowest_tests_count to that class and move this logic there.

File: tests/e2e/test_testing.py
Comment (kasperski95) [inline tests/e2e/test_testing.py:224]: Integration tests checking:
- the order of displaying the slowest tests
- if the amount of the slowest tests actually limits the output
would be nice.

(E2E tests are slow and should be use for testing the core functionality.)

File: protostar/commands/test/test_command.py
Comment (kasperski95) [inline protostar/commands/test/test_command.py:147]: We should generally avoid a first word that could be interpreted as a verb when naming variables:
```suggestion
        slowest_tests_to_report_count: int = 0,
```

**Review by kasperski95 — COMMENTED** [review]
Screenshot 2022-07-28 at 10 42 51

- I'd move the slowest tests section above [INFO] Test suites. In the console, the mose important information should be at the bottom.
- Green color should be reserved for representing success state (e.g. test pass). I'd change the color of "captured stdout" and color of the enumeration to the "CYAN" and remove the color from the enumeration
- For the enumeration, I'd use format `1.` instead of `[1]` and removed the tab between enumeration and a test case name
- Time should be in the same brackets with steps e.g. `(time=..., steps=...)`
- The slow test list item should have the same or similar look to the test case title (2.31 s) -> gray `(time=2.31s)`

File: protostar/commands/test/testing_summary.py
Comment (Arcticae) [inline protostar/commands/test/testing_summary.py:195]: let's just extract color to a variable instead of turning pylint off here

**Review by Arcticae — CHANGES_REQUESTED** [review]
Just a thought - could this be done as a boolean flag with optional parameter of number of slowest tests to log? We could get i.e. `min(tests_ran, 10)` as a default value

**Comment by kasperski95** [comment]
> Just a thought - could this be done as a boolean flag with optional parameter of number of slowest tests to log? We could get i.e. min(tests_ran, 10) as a default value

The same functionality can be achieved with one flag as it is now.

File: protostar/commands/test/testing_summary.py
Comment (kasperski95) [inline protostar/commands/test/testing_summary.py:187]: The name suggests returning a list of strings and that this function is a getter.

```suggestion
    def _format_slow_test_cases_list(self, count: int) -> str:
```

File: tests/integration/testing_timing/testing_timing_test.py
Comment (kasperski95) [inline tests/integration/testing_timing/testing_timing_test.py:12]:
```suggestion
def assert_execution_times_are_in_desc_order(logs: str) -> None:
```

File: tests/integration/testing_timing/testing_timing_test.py
Comment (kasperski95) [inline tests/integration/testing_timing/testing_timing_test.py:62]: Consider splitting asserts to test cases where `testing_summary` is passed as a fixture.

File: tests/integration/testing_timing/testing_timing_test.py
Comment (kasperski95) [inline tests/integration/testing_timing/testing_timing_test.py:58]:
```suggestion
    assert "4. " in test5
    assert "5. " not in test5 
```

File: protostar/commands/test/testing_summary.py
Comment (kasperski95) [inline protostar/commands/test/testing_summary.py:64]:
```suggestion
    def log_slowest_test_case(
```
Public methods should be more explicit. Somebody could think that this method logs THE slowest test case.

**Review by kasperski95 — CHANGES_REQUESTED** [review]
Try to be more explicit when naming functions and variables. Quite often I had to analyze the code to figure out what's going on instead of just reading the name.

Requesting changes because the integration test is too hard to read/maintain/fragile.

File: tests/integration/testing_timing/testing_timing_test.py
Comment (Radinyn) [reply tests/integration/testing_timing/testing_timing_test.py:58]: There are color issues with that approach + it's possible (though unlikely) that this will interfere with the test timing.
I believe that set comparison is more bullet-proof anyway.

File: protostar/commands/test/testing_summary.py
Comment (Radinyn) [reply protostar/commands/test/testing_summary.py]: Is it necessary to make it a dataclass? Most of the time when we deal with text like this (for example in ```test_cases.py```) we only use strings.

File: tests/integration/testing_timing/testing_timing_test.py
Comment (kasperski95) [reply tests/integration/testing_timing/testing_timing_test.py:67]:
```suggestion
    assert {1, 2, 3, 4} == get_first_digit_test_indices(test5)
```
I consider it unnecessarily complicated, but this is just a test. Your call.

File: tests/integration/testing_timing/testing_timing_test.py
Comment (kasperski95) [inline tests/integration/testing_timing/testing_timing_test.py:61]: `test_timing_...` -> `test_output_when_slowest_tests_count...`

File: protostar/commands/test/testing_summary.py
Comment (kasperski95) [reply protostar/commands/test/testing_summary.py]: It's not necessary. The end goal is to simplify the code. Maybe from this code a helper function `format_as_table` could be extracted.

I'll not block this PR if you leave this as it is though.

**Review by kasperski95 — APPROVED** [review]

File: protostar/commands/test/testing_summary.py
Comment (kasperski95) [reply protostar/commands/test/testing_summary.py]: It's not necessary. The end goal is to simplify the code. Maybe from this code a helper function `format_as_table` could be extracted.

I'll not block this PR if you leave this as it is though.

File: tests/e2e/test_testing.py
Comment (mkaput) [inline tests/e2e/test_testing.py:226]: you are not testing that the tests themselves have been printed out

File: protostar/commands/test/test_command.py
Comment (mkaput) [inline protostar/commands/test/test_command.py:116]: Why is this a configurable argument? IMHO people will not use this feature not even know that it exists, because it's disabled by default

File: tests/integration/testing_timing/testing_timing_test.py
Comment (mkaput) [inline tests/integration/testing_timing/testing_timing_test.py:15]: I think this assertion may be faulty if there are two tests with different names but exactly the same timings. This assertion may not check that tests are sorted by name as expected.

Why not simply asserting that the logged output matches a regular expression?

**Pull request merged**
